File: Day-15/main.ex
function cell_legit(integer x, integer y, integer cols, integer rows)
    if x >= 1 and x <= cols and y >= 1 and y <= rows then
        return true
    end if
    return false
end function


procedure main()
    object f = get_text("input.txt", GT_LF_STRIPPED)
    if f=-1 then
        puts(1, "Can't find input.txt file")
        abort(1)
    end if

    seq inp = {}
    for i=1 to length(f) do
        inp = append(inp, apply(split_by(f[i], 1), to_integer))
    end for

    int {rows, cols} @= length(inp)
    seq moves = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}}
    
    seq visited = repeat(repeat(0, rows), rows)
    pq_add({{1, 1}, 0})

    seq node
    int cost
    int {x, y} @= 0
    
    while pq_size() do
        {node, cost} = pq_pop()
        {x, y} = node
        if visited[x][y] != 1 then
            visited[x][y] = 1
            
            if x == rows and y == rows then
                printf(1, "Answer is: %d\n", cost)
                exit
            end if
            
            for i=1 to 4 do
                int {n_x, n_y} = {x + moves[i][1], y + moves[i][2]}
                if cell_legit(n_x, n_y, rows, rows) == true then
                    if visited[n_x][n_y] != 1 then
                        pq_add({{n_x, n_y}, cost + inp[n_x][n_y]})
                    end if
                end if
            end for
        end if
    end while
end procedure

main()

File: Day-15/main.py
from queue import PriorityQueue
from typing import List, Set, Tuple


board: List[List[int]] = []

with open('input.txt', 'r') as inp_f:
    for line in inp_f.readlines():
        board.append(list(map(int, line.strip())))

rows = len(board)
cols = len(board[0])
moves = ((-1, 0), (1, 0), (0, -1), (0, 1))

def cell_legit(cell: Tuple[int, int]) -> bool:
    return cols > cell[0] >= 0 and rows > cell[1] >= 0

def cell_neighbours(cell: Tuple[int, int]) -> List[Tuple[int, int]]:
    return [n_c for x, y in moves if cell_legit(n_c := (cell[0] + x, cell[1] + y))]

def ucs(graph: List[List[int]], start: Tuple[int, int], goal: Tuple[int, int]) -> int:
    visited: Set[Tuple[int, int]] = set()
    queue: PriorityQueue[Tuple[int, Tuple[int, int]]] = PriorityQueue()
    queue.put((0, start))

    while queue:
        cost, node = queue.get()
        if node not in visited:
            visited.add(node)

            if node == goal:
                return cost

            for i in cell_neighbours(node):
                if i not in visited:
                    queue.put((cost + graph[i[0]][i[1]], i))
    return -1


# Part 1
print(ucs(board, (0, 0), (rows - 1, cols - 1)))

def expand_matrix(matrix: List[List[int]], mag = 5) -> List[List[int]]:
    width: int = len(matrix[0])
    height: int = len(matrix)
    output = [[0 for _x in range(width*mag)] for _y in range(height*mag)]

    for y in range(mag):
        for x in range(mag):
            for b in range(height):
                for a in range(width):
                    output[(y*height) + b][(x*width) + a] = (matrix[b][a] + y + x - 1) % 9 + 1

    return output

# Part 2
board = expand_matrix(board)

rows = len(board)
cols = len(board[0])

print(ucs(board, (0, 0), (rows - 1, cols - 1)))
